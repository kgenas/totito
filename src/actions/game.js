import { types } from "../types/types";

export const startGame = () =>({
    type: types.start
})

export const Clicked = (cellValues,numberOfTurn,isNext,isGameOver,winner,winningCombination) =>({
    type: types.cellValues,
    payload:{
        cellValues,
        numberOfTurn,
        isNext,
        isGameOver,
        winner,
        winningCombination
    }  
})

export const NewGame = () =>({
    type: types.newGame
})