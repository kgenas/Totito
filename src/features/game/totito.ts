import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export interface InitialState {
    cellValues: string[];
    numberOfTurn: number;
    xIsNext: boolean;
}
const initialState: InitialState = {
    cellValues: ['','','','','','','','',''],
    numberOfTurn: 9,
    xIsNext: true,
};

export const game = createSlice({
  name: 'game',
  initialState,  
  reducers: {    
    click: (state, action: PayloadAction<InitialState>) => { 
      const { cellValues , numberOfTurn , xIsNext } = action.payload;
      return {
        ...state,
        cellValues,
        numberOfTurn,
        xIsNext
      };
    },    
    newGame: (state) => {
        state.cellValues = initialState.cellValues;
        state.numberOfTurn = initialState.numberOfTurn;
        state.xIsNext = initialState.xIsNext;
      },
  },
});

export const { click, newGame } = game.actions;
export default game.reducer;
