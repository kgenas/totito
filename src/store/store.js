
import { createStore, combineReducers } from 'redux';
import { game } from '../reducers/gameReducer';

const reducers = combineReducers({
    Game:game
});

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__();

export const store = createStore(
    reducers,
    composeEnhancers
);
