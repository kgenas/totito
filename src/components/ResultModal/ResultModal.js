import classNames from 'classnames';
import { useDispatch, useSelector } from 'react-redux';
import { NewGame } from '../../actions/game';
import './ResultModal.css';

export const ResultModal = () =>{

    const dispatch = useDispatch();
    const { isGameOver, winner } = useSelector( state => state.GameTotito) ;
    const resultMondalClasses = classNames({
        'modal-open': isGameOver
    })
    const message = winner ? `Winner is ${ winner }.` :'It is a tie';

    const newGame = () =>{
        dispatch(NewGame());
    }
    return(
        <div id='modal-overlay' className={ resultMondalClasses }>
            <div id='game-result-modal'>
                <div id='result-container'>
                    <div id='winner-container'>
                        <span>{ message }</span>
                    </div>
                </div>
                <div id='new-game-container'>
                    <button
                        id='new-game-container'
                        onClick={ newGame }
                    >
                        Start New Game
                    </button>
                </div>
            </div>            
        </div>
    )
}