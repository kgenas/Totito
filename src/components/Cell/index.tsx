import './Cell.css';
import classNames from 'classnames';
import { useSelector, useDispatch } from 'react-redux';
import { AppDispatch, RootState } from '../../app/store';
import { winnerCalculatorValue } from '../../features/game/result';
import { click } from '../../features/game/totito';

interface valueProps {    
    key: number ;
    value: string;
    valueKey: number;
    switchOnOff: boolean;
}
export const Cell = ( props: valueProps ) => {
          
    const { cellValues, numberOfTurn, xIsNext } = useSelector((state: RootState) => state.game );    
    const dispatch:AppDispatch = useDispatch();

    const isCellEmpty = ( cellIndex:number ) => cellValues[cellIndex] === '';
    const onCellClicked = () =>{ 
      
      if (isCellEmpty( props.valueKey )){
        const newCellValues = [...cellValues];
        newCellValues[props.valueKey] = xIsNext ? 'X':'O';
        dispatch(winnerCalculatorValue( newCellValues,(numberOfTurn-1),props.valueKey ));        
        dispatch(click(
            {
                cellValues: newCellValues,
                numberOfTurn: (numberOfTurn-1),
                xIsNext: !xIsNext,
            }
        ));
    }
    }
    const cellCss = classNames({
        cell: true,
        winner: props.switchOnOff
    })

    const cellContentCss = classNames({
        'cell--contet': true,
        populated: props.value 
    })

  return (
    <button
        className={ cellCss }
        onClick = { onCellClicked }
    >
        <span
            className={ cellContentCss }
        >
            { props.value }
        </span>
    </button>
  )
}
