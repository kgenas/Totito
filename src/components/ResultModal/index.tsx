import classNames from 'classnames';
import { useSelector, useDispatch } from 'react-redux';
import { newGame } from '../../features/game/totito';
import './ResultModal.css';
import { AppDispatch, RootState } from '../../app/store';
import { winnerCalculatorStart } from '../../features/game/result';

export const ResultModal = () =>{

    const dispatch:AppDispatch = useDispatch();
    const { winner , isGameOver } = useSelector((state: RootState) => state.result );    

    const resultMondalClasses = classNames({
        'modal-overlay': true,
        'modal-overlay--modal-open': isGameOver

    })
    const message = winner ? `Winner is ${ winner }.` :'It is a tie';

    const newGames = () =>{
        dispatch(newGame());
        dispatch(winnerCalculatorStart());
    }
    return(
        <div className= { resultMondalClasses }>
            <div className='game-result-modal'>
                <div className='result-container'>
                    <div className='winner-container'>
                        <span>{ message }</span>
                    </div>
                </div>
                <div className='new-game-container'>
                    <button
                        className='new-game-container__button'
                        onClick={ newGames }
                    >
                        Start New Game
                    </button>
                </div>
            </div>            
        </div>
    )
}