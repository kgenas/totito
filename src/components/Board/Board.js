import './Board.css';
import { useSelector } from 'react-redux';
import { Cell } from '../Cell/Cell';

export const Board = (props) => {
  
  const { cellValues, winningCombination } = useSelector( state => state.Game)
  const cell = cellValues.map((value, index)=>{
        
    const switchOnOff = winningCombination && winningCombination.indexOf(index) >= 0;
    return(
        <Cell
            key= { index }
            value = { value }
            switchOnOff = { switchOnOff }
            onClick = { () => props.cellClicked(index) }
        />
    )
  })

  return (
    <div className='board'>
        { cell }
    </div>
  )
}
