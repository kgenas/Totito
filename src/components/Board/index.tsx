import './Board.css';
import { useSelector } from 'react-redux';
import { Cell } from '../Cell';
import { RootState } from '../../app/store';
export const Board = () => {
    
  const { cellValues }  = useSelector((state: RootState) => state.game );
  const { winningCombination } = useSelector((state: RootState) => state.result ); 

  const cell = cellValues.map((value:string, index:number)=>{
       
  const switchOnOff = winningCombination && winningCombination.indexOf( index ) >= 0;
    return(
        <Cell
            key = { index }
            value = { value }
            valueKey = { index }
            switchOnOff = { switchOnOff }
        />
    )
  })

  return (
    <div className='board'>
        { cell }
    </div>
  )
}
