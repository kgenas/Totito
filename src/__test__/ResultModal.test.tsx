/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable testing-library/no-debugging-utils */
import React from 'react';
import { screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { ResultModal } from '../components/ResultModal';
import { renderWithProviders } from '../utils/test-utils';
import { setupStore } from '../app/store';
import { click, newGame } from '../features/game/totito';
import { Board } from '../components/Board';

describe('ResultModal',() =>{
    // Note: running cleanup afterEach is done automatically for you in @testing-library/react@9.0.0 or higher
    // unmount and cleanup DOM after the test is finished.
    //? afterEach(cleanup);
    test('Renders ResultModal component',() => {
        renderWithProviders(<ResultModal/>);
        screen.debug();
    });

    test('Renders button',() => {
        renderWithProviders(<ResultModal/>);
        const value = screen.getByText(/start new game/i);
        expect(value).toBeInTheDocument();
    });

    test('Matches snapshot', () => { 

        const { baseElement } = renderWithProviders(<ResultModal/>);
        expect( baseElement ).toMatchSnapshot();
    });

    test('New Game', () => {        
        const store = setupStore();
        const newCellValues:string[] = ['X','','X','O','','O','X','',''];
        store.dispatch(click({
                    cellValues:newCellValues,
                    numberOfTurn: (9-1),
                    xIsNext: false,
        }))        
        renderWithProviders(<Board/>, { store });        
        //* can we update Board
        act( () => {
            store.dispatch(newGame());
        });       
        renderWithProviders(<Board/>, { store });

        const value = screen.getAllByRole('button');
        expect(value[0]).toHaveTextContent('');
        screen.debug();
    });
    
});
