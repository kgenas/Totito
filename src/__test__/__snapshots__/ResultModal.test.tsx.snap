// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultModal Matches snapshot 1`] = `
<body>
  <div>
    <div
      class="modal-overlay"
    >
      <div
        class="game-result-modal"
      >
        <div
          class="result-container"
        >
          <div
            class="winner-container"
          >
            <span>
              It is a tie
            </span>
          </div>
        </div>
        <div
          class="new-game-container"
        >
          <button
            class="new-game-container__button"
          >
            Start New Game
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
